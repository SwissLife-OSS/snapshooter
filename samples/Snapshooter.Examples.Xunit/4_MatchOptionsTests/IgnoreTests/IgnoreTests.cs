using Snapshooter.Tests.Data;
using Snapshooter.Xunit;
using System;
using Xunit;

namespace Snapshooter.Quickstarts.Xunit
{
    public class IgnoreTests
    {
        /// <summary>
        /// Tests if the new created person is valid. The id will always be newly created,
        /// therefore we have to ignore the id of the person during comparison.
        /// The path to the id is given by a json path.
        /// </summary>
        [Fact]
        public void CreatePerson_VerifyPersonByIgnoringId_SuccessfulVerified()
        {
            // arrange
            var serviceClient = new ServiceClient();

            // act
            TestPerson person = serviceClient.CreatePerson(
                Guid.NewGuid(), "David", "Mustermann");

            // assert
            Snapshot.Match(person, matchOptions => matchOptions.IgnoreField("Id"));
        }

        /// <summary>
        /// Tests if the new created person is valid. The id will always be newly created,
        /// therefore we have to ignore the id of the person during comparison.
        /// The path to the id is given by a json path.
        /// In Xunit, the Theory InlineData parameters cannot be read, therefore the parameters
        /// can be added to the snapshot name with the snapshot name extension functionality.
        /// </summary>
        [Theory]
        [InlineData("David", "Mustermann")]
        [InlineData("Claudia", "Musterfrau")]

        public void CreatePerson_VerifyPersonsByIgnoringId_SuccessfulVerified(
            string firstname, string lastname)
        {
            // arrange
            var serviceClient = new ServiceClient();

            // act
            TestPerson person = serviceClient.CreatePerson(
                Guid.NewGuid(), firstname, lastname);

            // assert
            Snapshot.Match(person, 
                SnapshotNameExtension.Create(firstname, lastname), 
                matchOptions => matchOptions.IgnoreField("Id"));
        }

        /// <summary>
        /// Tests if the new created person is valid. If we want to ignore several
        /// fields of the result object, we can concat the ignore options as many
        /// times as we want.
        /// The path to the id is given by a json path.
        /// </summary>
        [Fact]
        public void CreatePerson_VerifyPersonByIgnoringSeveralFields_SuccessfulVerified()
        {
            // arrange
            var serviceClient = new ServiceClient();

            // act
            TestPerson person = serviceClient.CreatePerson(
                Guid.NewGuid(), "David", "Mustermann");

            // assert
            Snapshot.Match(person, 
                matchOptions => matchOptions
                    .IgnoreField("Id")
                    .IgnoreField("Address.Plz")
                    .IgnoreField("Relatives[0].Address.Country.Name"));
        }

        /// <summary>
        /// Tests if the new created person is valid. We ignore all
        /// fields of the name 'Id' from the object.
        /// The path to the id is given by a json path.
        /// </summary>
        [Fact]
        public void CreatePerson_VerifyPersonByIgnoringAllFieldsOfASpecificName_SuccessfulVerified()
        {
            // arrange
            TestPerson person = TestDataBuilder
                .TestPersonMarkWalton()
                .Build();

            // act
            person.Id = Guid.NewGuid();
            person.Relatives[0].Id = Guid.NewGuid();

            // assert
            Snapshot.Match(person, matchOptions => matchOptions.IgnoreField<Guid>("**.Id"));
        }

        /// <summary>
        /// Tests if the new created person is valid. We ignore all
        /// fields of the name 'Id' from the object.
        /// The path to the id is given by a json path.
        /// </summary>
        [Fact]
        public void CreatePerson_VerifyPersonByIgnoringAllIdFields_SuccessfulVerified()
        {
            // arrange
            TestPerson person = TestDataBuilder
                .TestPersonMarkWalton()
                .Build();

            // act
            person.Id = Guid.NewGuid();
            person.Relatives[0].Id = Guid.NewGuid();

            // assert
            Snapshot.Match(person, matchOptions => matchOptions.IgnoreAllFields<Guid>("Id"));
        }
    }
}
