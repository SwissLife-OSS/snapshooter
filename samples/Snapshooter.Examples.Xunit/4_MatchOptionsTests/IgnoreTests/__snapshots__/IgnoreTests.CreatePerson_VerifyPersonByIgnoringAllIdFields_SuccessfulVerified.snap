{
  "Id": "5e5b8485-a6a9-4c2b-937a-f0117b7874b2",
  "Firstname": "Mark",
  "Lastname": "Walton",
  "CreationDate": "2018-06-06T00:00:00",
  "DateOfBirth": "2000-06-25T00:00:00",
  "Age": 30,
  "Size": 182.5214,
  "Address": {
    "Street": "Rohrstrasse",
    "StreetNumber": 12,
    "Plz": 8304,
    "City": "Wallislellen",
    "Country": {
      "Name": "Switzerland",
      "CountryCode": "CH"
    }
  },
  "Children": [
    {
      "Name": "James",
      "DateOfBirth": "2015-02-12T00:00:00"
    },
    {
      "Name": null,
      "DateOfBirth": "2015-02-12T00:00:00"
    },
    {
      "Name": "Hanna",
      "DateOfBirth": "2012-03-20T00:00:00"
    }
  ],
  "Relatives": [
    {
      "Id": "8d682d55-4573-4aa0-a6d6-f15d5e3b127f",
      "Firstname": "Sandra",
      "Lastname": "Schneider",
      "CreationDate": "2019-04-01T00:00:00",
      "DateOfBirth": "1996-02-14T00:00:00",
      "Age": null,
      "Size": 165.23,
      "Address": {
        "Street": "Bahnhofstrasse",
        "StreetNumber": 450,
        "Plz": 8000,
        "City": "Zurich",
        "Country": {
          "Name": "Switzerland",
          "CountryCode": "CH"
        }
      },
      "Children": [],
      "Relatives": null
    }
  ]
}
